File: LeetCode/Java/DeleteNodeInaLinkedList.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class DeleteNodeInaLinkedList {
    
    public void deleteNode(ListNode node) {
        if (node == null || node.next == null) {
            return;
        }
        node.val = node.next.val;
        node.next = node.next.next;
    }

    public static void printList(ListNode head) {
        ListNode current = head;
        while (current != null) {
            System.out.print(current.val);
            if (current.next != null) {
                System.out.print(" -> ");
            }
            current = current.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(4);
        head.next = new ListNode(5);
        head.next.next = new ListNode(1);
        head.next.next.next = new ListNode(9);

        System.out.print("Original List: ");
        printList(head);

        ListNode nodeToDelete = head.next;

        DeleteNodeInaLinkedList obj = new DeleteNodeInaLinkedList();
        obj.deleteNode(nodeToDelete);

        System.out.print("After Deleting 5: ");
        printList(head);
    }
}


// Input: head = [4,5,1,9], node = 5
// Output: [4,1,9]

// Input: head = [4,5,1,9], node = 1
// Output: [4,5,9]

File: LeetCode/Python/DeleteNodeinaLinkedList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, value):
        self.value = value
        self.next = None

class Solution:
    def deleteNode(self, node):
        node.value = node.next.value
        node.next = node.next.next


def constructLinkedList(ls):
    head = ListNode(ls[0])
    temp = head
    nodes = [head]   # to keep references
    for i in ls[1:]:
        temp.next = ListNode(i)
        temp = temp.next
        nodes.append(temp)
    return head, nodes


def printLinkedList(head):
    temp = head
    while temp:
        print(temp.value, end=' -> ')
        temp = temp.next
    print("None")


# Driver Code
head_list = [4,5,1,9]
head, nodes = constructLinkedList(head_list)

# We want to delete node with value 5
node_to_delete = nodes[1]   # This is actual ListNode, not integer

obj = Solution()

print("Before deleting:")
printLinkedList(head)

obj.deleteNode(node_to_delete)

print("After deleting:")
printLinkedList(head)